pyanaconda/dbus/typing.py

```diff
@@ -0,0 +1,182 @@
+Int64 = NewType('Int64', int)
+UInt64 = NewType('UInt64', int)
+
+# Type of a file handler.
+File = IO
+
+# Type of an object path.
+ObjPath = NewType('ObjPath', str)
+
+# Container types.
+# Use Tuple, Dict and List from typing.
+
+# Variant type.
+Variant = Union[
+    Bool,
+    Double,
+    Str,
+    Int,
+    Byte,
+    Int16,
+    UInt16,
+    Int32,
+    UInt32,
+    Int64,
+    UInt64,
+    File,
+    ObjPath,
+    List[Any],
+    Dict[Any, Any],
+    Tuple[Any, ...]
+]
+
+def get_dbus_type(type_hint):
+    """Return DBus representation of a type hint.
+
+    :param type_hint: a type hint
+    :return: a string with DBus representation
+    """
+    return DBusType.get_dbus_representation(type_hint)
+
+
+class DBusType(object):
+    """Class for transforming type hints to DBus types."""
+
+    # DBus representation of basic types.
+    _basic_type_mapping = {
+        # Basic types.
+        Bool:       "b",
+        Str:        "s",
+        Double:     "d",
+        # Default integer.
+        Int:        "i",
+        # Integer types.
+        Byte:       "y",
+        Int16:      "n",
+        UInt16:     "q",
+        Int32:      "i",
+        UInt32:     "u",
+        Int64:      "x",
+        UInt64:     "t",
+        # Other basic types.
+        File:       "h",
+        ObjPath:    "o",
+        Variant:    "v"
+    }
+
+    # DBus representation of container types.
+    _container_type_mapping = {
+        Tuple:      "(%s)",
+        List:       "a%s",
+        Dict:       "a{%s}",
+    }
+
+    @staticmethod
+    def _is_basic_type(type_hint):
+        """Is is a basic type?"""
+        return type_hint in DBusType._basic_type_mapping
+
+    @staticmethod
+    def _get_basic_type(type_hint):
+        """Return a basic type."""
+        return DBusType._basic_type_mapping[type_hint]
+
+    @staticmethod
+    def _is_container_type(type_hint):
+        """Is it a container type?"""
+        # Try to get the "base" type of the container type.
+        origin = getattr(type_hint, "__origin__", None)
+        return origin in DBusType._container_type_mapping
+
+    @staticmethod
+    def _get_container_type(type_hint):
+        """Return a container type."""
+        # Get the "base" type of the container.
+        origin = type_hint.__origin__
+        # Get the arguments of the container.
+        args = type_hint.__args__
+
+        # Check the typing.
+        if origin == Dict:
+            DBusType._check_dictionary(type_hint)
+
+        # Generate string.
+        container = DBusType._container_type_mapping[origin]
+        items = [DBusType.get_dbus_representation(arg) for arg in args]
+        return container % "".join(items)
+
+    @staticmethod
+    def _check_dictionary(type_hint):
+        """Check the type of a dictionary.
+
+        Raises ValueError, if the type is not valid.
+        """
+        key, _ = type_hint.__args__
+
+        if DBusType._is_container_type(key) or key == Variant:
```

Comment:
DBus specification doesn't allow it:
> A DICT_ENTRY works exactly like a struct, but rather than parentheses it uses curly braces, and it has more restrictions. The restrictions are: it occurs only as an array element type; it has exactly two single complete types inside the curly braces; the first single complete type (the "key") must be a basic type rather than a container type. Implementations must not accept dict entries outside of arrays, must not accept dict entries with zero, one, or more than two fields, and must not accept dict entries with non-basic-typed keys. A dict entry is always a key-value pair. 

https://dbus.freedesktop.org/doc/dbus-specification.html#type-system